src/Common/ColumnsHashing.h

```diff
@@ -383,37 +384,72 @@ struct HashMethodSerialized
             for (auto row_size : row_sizes)
                 total_size += row_size;
 
-            serialized_buffer.resize(total_size);
-
-            const size_t rows = row_sizes.size();
-            char * memory = serialized_buffer.data();
-            std::vector<char *> memories(rows);
-            serialized_keys.resize(rows);
-            for (size_t i = 0; i < row_sizes.size(); ++i)
+            use_batch_serialize = shouldUseBatchSerialize();
+            if (use_batch_serialize)
             {
-                memories[i] = memory;
-                serialized_keys[i].data = memory;
-                serialized_keys[i].size = row_sizes[i];
+                serialized_buffer.resize(total_size);
 
-                memory += row_sizes[i];
-            }
+                const size_t rows = row_sizes.size();
+                char * memory = serialized_buffer.data();
+                std::vector<char *> memories(rows);
+                serialized_keys.resize(rows);
+                for (size_t i = 0; i < row_sizes.size(); ++i)
+                {
+                    memories[i] = memory;
+                    serialized_keys[i].data = memory;
+                    serialized_keys[i].size = row_sizes[i];
 
-            for (size_t i = 0; i < keys_size; ++i)
-            {
-                if constexpr (nullable)
-                    key_columns[i]->batchSerializeValueIntoMemoryWithNull(memories, null_maps[i]);
-                else
-                    key_columns[i]->batchSerializeValueIntoMemory(memories);
+                    memory += row_sizes[i];
+                }
+
+                for (size_t i = 0; i < keys_size; ++i)
+                {
+                    if constexpr (nullable)
+                        key_columns[i]->batchSerializeValueIntoMemoryWithNull(memories, null_maps[i]);
+                    else
+                        key_columns[i]->batchSerializeValueIntoMemory(memories);
+                }
             }
         }
     }
 
+    bool shouldUseBatchSerialize() const
+    {
+#if defined(__aarch64__)
+        return true;
+#endif
+
+        size_t l2_size = 256 * 1024;
+#if defined(OS_LINUX) && defined(_SC_LEVEL2_CACHE_SIZE)
+        if (auto ret = sysconf(_SC_LEVEL2_CACHE_SIZE); ret != -1)
+            l2_size = ret;
+#endif
+        size_t avg_row_size = total_size / std::max(row_sizes.size(), 1UL);
+        return total_size <= 4 * l2_size && avg_row_size < 128;
+    }
+
     friend class columns_hashing_impl::HashMethodBase<Self, Value, Mapped, false>;
 
     ALWAYS_INLINE ArenaKeyHolder getKeyHolder(size_t row, Arena & pool) const
     requires(prealloc)
     {
-        return ArenaKeyHolder{serialized_keys[row], pool};
+        if (use_batch_serialize)
+            return ArenaKeyHolder{serialized_keys[row], pool};
+        else
+        {
+            std::unique_ptr<char[]> holder = std::make_unique<char[]>(row_sizes[row]);
```

Comment:
The allocation should use the arena pool instead. IIUC, the fallback logic can simply reuse the original per-row preallocation code.